### ISSUE TITLE
ICE: index out of bounds when trying to use associated const in array size

### ISSUE THREAD

**Issue by edsrzf** [rust-lang/rust#76122]
### Code

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e531c045b6135d1ddfd64074feaaab5d)

```Rust
trait Trait {
    type Type;
    
    const VALUE: Self::Type;
}

trait TraitExt {
    type Array;
}

impl<T> TraitExt for dyn Trait<Type=T> {
    type Array = [usize; Self::VALUE];
}
```

(I now realize that the thing I was trying to do was misguided and wrong, but hey, an ICE is an ICE.)

### Meta
`rustc --version --verbose`:
```
rustc 1.46.0 (04488afe3 2020-08-24)
binary: rustc
commit-hash: 04488afe34512aa4c33566eb16d8c912a3ae04f9
commit-date: 2020-08-24
host: x86_64-apple-darwin
release: 1.46.0
LLVM version: 10.0
```

The bug also exists on nightly as of the time of this report.

### Error output

```
thread 'rustc' panicked at 'index out of bounds: the len is 0 but the index is 0', src/librustc_middle/ty/mod.rs:895:14
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
   1: core::fmt::write
   2: std::io::Write::write_fmt
   3: std::panicking::default_hook::{{closure}}
   4: std::panicking::default_hook
   5: rustc_driver::report_ice
   6: std::panicking::rust_panic_with_hook
   7: rust_begin_unwind
   8: core::panicking::panic_fmt
   9: core::panicking::panic_bounds_check
  10: rustc_middle::ty::Generics::param_at
  11: rustc_middle::ty::error::<impl rustc_middle::ty::context::TyCtxt>::note_and_explain_type_err
  12: rustc_infer::infer::error_reporting::<impl rustc_infer::infer::InferCtxt>::note_type_err
  13: rustc_infer::infer::error_reporting::<impl rustc_infer::infer::InferCtxt>::report_and_explain_type_error
  14: rustc_infer::infer::InferCtxt::report_mismatched_types
  15: rustc_typeck::check::demand::<impl rustc_typeck::check::FnCtxt>::demand_coerce_diag
  16: rustc_typeck::check::expr::<impl rustc_typeck::check::FnCtxt>::check_expr_coercable_to_type
  17: rustc_infer::infer::InferCtxtBuilder::enter
  18: rustc_typeck::check::typeck_tables_of
  19: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::typeck_tables_of>::compute
  20: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  21: rustc_data_structures::stack::ensure_sufficient_stack
  22: rustc_query_system::query::plumbing::get_query_impl
  23: rustc_mir::const_eval::eval_queries::const_eval_raw_provider
  24: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::const_eval_raw>::compute
  25: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  26: rustc_data_structures::stack::ensure_sufficient_stack
  27: rustc_query_system::query::plumbing::get_query_impl
  28: rustc_mir::const_eval::eval_queries::const_eval_validated_provider
  29: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::const_eval_validated>::compute
  30: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  31: rustc_data_structures::stack::ensure_sufficient_stack
  32: rustc_query_system::query::plumbing::get_query_impl
  33: rustc_mir::const_eval::eval_queries::const_eval_validated_provider
  34: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::const_eval_validated>::compute
  35: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  36: rustc_data_structures::stack::ensure_sufficient_stack
  37: rustc_query_system::query::plumbing::get_query_impl
  38: rustc_middle::mir::interpret::queries::<impl rustc_middle::ty::context::TyCtxt>::const_eval_global_id
  39: rustc_middle::mir::interpret::queries::<impl rustc_middle::ty::context::TyCtxt>::const_eval_resolve
  40: <rustc_trait_selection::traits::project::AssocTypeNormalizer as rustc_middle::ty::fold::TypeFolder>::fold_const
  41: <rustc_trait_selection::traits::project::AssocTypeNormalizer as rustc_middle::ty::fold::TypeFolder>::fold_ty
  42: rustc_trait_selection::traits::project::opt_normalize_projection_type
  43: rustc_trait_selection::traits::project::normalize_projection_type
  44: <rustc_trait_selection::traits::project::AssocTypeNormalizer as rustc_middle::ty::fold::TypeFolder>::fold_ty
  45: rustc_middle::ty::fold::TypeFoldable::fold_with
  46: rustc_middle::ty::fold::TypeFoldable::fold_with
  47: rustc_trait_selection::traits::project::AssocTypeNormalizer::fold
  48: rustc_trait_selection::traits::project::normalize_to
  49: rustc_trait_selection::traits::wf::WfPredicates::normalize
  50: rustc_trait_selection::traits::wf::trait_obligations
  51: rustc_infer::infer::InferCtxtBuilder::enter
  52: rustc_typeck::check::wfcheck::check_item_well_formed
  53: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::check_item_well_formed>::compute
  54: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  55: rustc_data_structures::stack::ensure_sufficient_stack
  56: rustc_query_system::query::plumbing::get_query_impl
  57: rustc_query_system::query::plumbing::ensure_query_impl
  58: <rustc_typeck::check::wfcheck::CheckTypeWellFormedVisitor as rustc_hir::itemlikevisit::ParItemLikeVisitor>::visit_item
  59: rustc_data_structures::sync::par_for_each_in
  60: rustc_hir::hir::Crate::par_visit_all_item_likes
  61: rustc_session::session::Session::track_errors
  62: rustc_typeck::check_crate
  63: rustc_interface::passes::analysis
  64: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::analysis>::compute
  65: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  66: rustc_query_system::query::plumbing::get_query_impl
  67: rustc_middle::ty::context::tls::enter_global
  68: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  69: rustc_span::with_source_map
  70: rustc_interface::interface::create_compiler_and_run
  71: scoped_tls::ScopedKey<T>::set
```

</p>
</details>

**Comment by jonas-schievink** [comment]
Duplicate of https://github.com/rust-lang/rust/issues/43408
